onnxruntime/core/graph/contrib_ops/nhwc_inference_context.cc

```diff
@@ -0,0 +1,375 @@
+  // copy any existing type and shape info, and convert to NCHW for usage with the ONNX inferencing
+  const auto* nhwc_type = ctx_.getInputType(0);
+  if (nhwc_type != nullptr) {
+    TransposeShapeToNchw(*nhwc_type, input_type_);
+  }
+
+  nhwc_type = ctx_.getOutputType(0);
+  if (nhwc_type != nullptr) {
+    TransposeShapeToNchw(*nhwc_type, output_type_);
+  }
+}
+
+const ONNX_NAMESPACE::AttributeProto* NhwcInferenceContext::getAttribute(const std::string& name) const {
+  return ctx_.getAttribute(name);
+}
+
+size_t NhwcInferenceContext::getNumInputs() const noexcept {
+  return ctx_.getNumInputs();
+}
+
+const ONNX_NAMESPACE::TypeProto* NhwcInferenceContext::getInputType(size_t index) const {
+  return (index == 0) ? &input_type_ : ctx_.getInputType(index);
+}
+
+const ONNX_NAMESPACE::TensorProto* NhwcInferenceContext::getInputData(size_t index) const {
+  // we can't return the NHWC input data without transposing it, but wouldn't expect to be asked for it
+  // during shape inferencing as getInputData is only used to retrieve things that may have small
+  // constant initializers (e.g. something like the min and max values of a Clip operator).
+  return index == 0 ? nullptr : ctx_.getInputData(index);
+}
+
+size_t NhwcInferenceContext::getNumOutputs() const noexcept {
+  return ctx_.getNumOutputs();
+}
+
+ONNX_NAMESPACE::TypeProto* NhwcInferenceContext::getOutputType(size_t index) {
+  return (index == 0) ? &output_type_ : ctx_.getOutputType(index);
+}
+
+ONNX_NAMESPACE::GraphInferencer* NhwcInferenceContext::getGraphAttributeInferencer(const std::string& attribute_name) {
+  return ctx_.getGraphAttributeInferencer(attribute_name);
+}
+
+const ONNX_NAMESPACE::TensorShapeProto* NhwcInferenceContext::getSymbolicInput(size_t index) const {
+  return ctx_.getSymbolicInput(index);
+}
+
+const ONNX_NAMESPACE::SparseTensorProto* NhwcInferenceContext::getInputSparseData(size_t index) const {
+  return ctx_.getInputSparseData(index);
+}
+
+// Propagate the inferred type/shape info to output 0, converting any inferred shape from NCHW to NHWC.
+void NhwcInferenceContext::PropagateOutputShape() {
+  auto& nhwc_tp = *ctx_.getOutputType(0);
+  TransposeShapeToNhwc(output_type_, nhwc_tp);
+}
+
+//
+// NhwcResizeInferenceContext
+//
+
+NhwcResizeInferenceContext::NhwcResizeInferenceContext(ONNX_NAMESPACE::InferenceContext& ctx) : ctx_(ctx) {
+  // copy any existing type and shape info, and convert to NCHW for usage with the ONNX inferencing
+  const auto* nhwc_type = ctx_.getInputType(0);
+  if (nhwc_type != nullptr) {
+    TransposeShapeToNchw(*nhwc_type, input_type_);
+  }
+
+  const size_t num_inputs = ctx_.getNumInputs();
+
+  // Skip ROI input processing.
+  // Enable processing ROI when an EP which supports ROI starts using layout transformer.
+  // NNAPI and QNN, which currently use the layout transformer, do not support it.
+
+  // Scales input
+  if (num_inputs > scales_input_index) {
+    TransposeScalesInput();
```

Comment:
Thank you for your help. This stuff can be hairy, especially when one is not familiar with this side of the code base.

As far as I understand, the `TransformLayoutForEP ` function already has a special case for NHWC Resize that correctly transposes all inputs to NHWC.

https://github.com/microsoft/onnxruntime/blob/d76cf374c48f1849193cc958e14f1851a723cb20/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc#L954-L979

When I dump the model during the 3 phases of layout transformation, I can see that the `scales` and `sizes` inputs are correctly transposed to NHWC. However, ONNX shape inferencing still expects these inputs to be in NCHW layout in order to correctly infer the output's shape. Otherwise, the onnx inferred output shape will not match the expected shape, which triggers this warning/error:

https://github.com/microsoft/onnxruntime/blob/d76cf374c48f1849193cc958e14f1851a723cb20/onnxruntime/core/graph/graph.cc#L100-L111

This `NhwcResizeInferenceContext` adapter modifies a _copy_ of the `scales` and `sizes` inputs to NCHW and presents that to shape inferencing. To me, the fact that we don't have restore the values of the `scales` and `sizes` inputs is further indication that the original graph is not being modified. Although, I could easily be mistaken.

In any case, I agree that having to do this does not feel entirely correct.